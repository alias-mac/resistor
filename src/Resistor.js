/*
 * Copyright (c) 2016 Filipe Guerra
 * https://github.com/alias-mac/resistor
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */

const significantDigits = [
  { value: 0, color: 'black' },
  { value: 1, color: 'brown' },
  { value: 2, color: 'red' },
  { value: 3, color: 'orange' },
  { value: 4, color: 'yellow' },
  { value: 5, color: 'green' },
  { value: 6, color: 'blue' },
  { value: 7, color: 'violet' },
  { value: 8, color: 'gray' },
  { value: 9, color: 'white' },
];

const multipliers = [
  { value: 0, color: 'black' },
  { value: 1, color: 'brown' },
  { value: 2, color: 'red' },
  { value: 3, color: 'orange' },
  { value: 4, color: 'yellow' },
  { value: 5, color: 'green' },
  { value: 6, color: 'blue' },
  { value: 7, color: 'violet' },
  { value: 8, color: 'gray' },
  { value: 9, color: 'white' },
  { value: -1, color: 'gold' },
  { value: -2, color: 'silver' },
];

const tolerances = [
  { value: 1, color: 'brown' },
  { value: 2, color: 'red' },
  { value: 0.5, color: 'green' },
  { value: 0.25, color: 'blue' },
  { value: 0.1, color: 'violet' },
  { value: 5, color: 'gold' },
  { value: 10, color: 'silver' },
  // {value: 20, color: 'none'},
];

const tempCoefs = [
  { value: 250, color: 'black' },
  { value: 100, color: 'brown' },
  { value: 50, color: 'red' },
  { value: 15, color: 'orange' },
  { value: 25, color: 'yellow' },
  { value: 20, color: 'green' },
  { value: 10, color: 'blue' },
  { value: 5, color: 'violet' },
  { value: 1, color: 'gray' },
];

function sanitize(resistor) {

  let attrs = resistor.attributes;

  [
    'bands',
    'digit1',
    'digit2',
    'digit3',
    'multiplier',
  ].forEach((key) => {
    attrs[key] = parseInt(attrs[key], 10) || 0;
  });

  attrs.tolerance = parseFloat(attrs.tolerance) || 10;
  attrs.tempCoef = parseInt(attrs.tempCoef, 10) || 250;
}

class Resistor {

  constructor(attributes) {
    this.attributes = {};
    this.set(attributes);
  }

  get(key) {
    return key ? this.attributes[key] : this.attributes;
  }

  set(key, value) {
    let attrs;
    if (typeof key === 'object') {
      attrs = key;
    } else {
      (attrs = {})[key] = value;
    }

    for (let attr in attrs) {
      value = attrs[attr];
      this.attributes[attr] = value;
    }

    sanitize(this);

    return this;
  }

  toOhms() {

    let [d1, d2, d3, m] = [
      this.get('digit1'),
      this.get('digit2'),
      this.get('digit3'),
      this.get('multiplier')
    ].map((int) => parseInt(int, 10) || 0);

    switch (this.get('bands')) {
      case 3:
      case 4:
        d1 *= 10;
        break;
      case 5:
      case 6:
        d1 *= 100;
        d2 *= 10;
        break;
    }
    let ohmage = d1 + d2 + d3;

    return ohmage * Math.pow(10, m);
  }

  static get significantDigits() {
    return significantDigits;
  }

  static get multipliers() {
    return multipliers;
  }

  static get tolerances() {
    return tolerances;
  }

  static get tempCoefs() {
    return tempCoefs;
  }
}

export default Resistor;
