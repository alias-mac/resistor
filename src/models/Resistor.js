/*
 * Copyright (c) 2016 Filipe Guerra
 * https://github.com/alias-mac/resistor
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */

import _find from 'lodash/find';

const significantDigits = [
  { value: 0, color: 'black' },
  { value: 1, color: 'brown' },
  { value: 2, color: 'red' },
  { value: 3, color: 'orange' },
  { value: 4, color: 'yellow' },
  { value: 5, color: 'green' },
  { value: 6, color: 'blue' },
  { value: 7, color: 'violet' },
  { value: 8, color: 'gray' },
  { value: 9, color: 'white' },
];

const multipliers = [
  { value: 0, color: 'black' },
  { value: 1, color: 'brown' },
  { value: 2, color: 'red' },
  { value: 3, color: 'orange' },
  { value: 4, color: 'yellow' },
  { value: 5, color: 'green' },
  { value: 6, color: 'blue' },
  { value: 7, color: 'violet' },
  { value: 8, color: 'gray' },
  { value: 9, color: 'white' },
  { value: -1, color: 'gold' },
  { value: -2, color: 'silver' },
];

const tolerances = [
  { value: 1, color: 'brown' },
  { value: 2, color: 'red' },
  { value: 0.5, color: 'green' },
  { value: 0.25, color: 'blue' },
  { value: 0.1, color: 'violet' },
  { value: 5, color: 'gold' },
  { value: 10, color: 'silver' },
  // {value: 20, color: 'none'},
];

const tempCoefs = [
  { value: 250, color: 'black' },
  { value: 100, color: 'brown' },
  { value: 50, color: 'red' },
  { value: 15, color: 'orange' },
  { value: 25, color: 'yellow' },
  { value: 20, color: 'green' },
  { value: 10, color: 'blue' },
  { value: 5, color: 'violet' },
  { value: 1, color: 'gray' },
];

function sanitize(resistor) {
  const attrs = resistor.attributes;

  [
    'bands',
    'digit1',
    'digit2',
    'digit3',
    'multiplier',
  ].forEach((key) => {
    attrs[key] = parseInt(attrs[key], 10) || 0;
  });

  attrs.tolerance = parseFloat(attrs.tolerance) || 10;
  attrs.tempCoef = parseInt(attrs.tempCoef, 10) || 250;
}

class Resistor {

  constructor(attributes) {
    this.attributes = {};
    this.set(attributes);
  }

  get(key) {
    return key ? this.attributes[key] : this.attributes;
  }

  set(key, value) {
    let attrs;
    if (typeof key === 'object') {
      attrs = key;
    } else {
      (attrs = {})[key] = value;
    }

    Object.keys(attrs).forEach((attr) => {
      this.attributes[attr] = attrs[attr];
    });

    sanitize(this);

    return this;
  }

  toOhms() {

    let [d1, d2, d3, m] = [ // eslint-disable-line prefer-const
      this.get('digit1'),
      this.get('digit2'),
      this.get('digit3'),
      this.get('multiplier'),
    ].map(int => parseInt(int, 10) || 0);

    switch (this.get('bands')) { // eslint-disable-line default-case
      case 3:
      case 4:
        d1 *= 10;
        break;
      case 5:
      case 6:
        d1 *= 100;
        d2 *= 10;
        break;
    }

    const ohmage = d1 + d2 + d3;

    return ohmage * (10 ** m); // Ω * 10^m
  }

  toString(options = {}) {

    const short = options.short !== undefined ? options.short : true;

    let result = this.toOhms();

    const unit = short ? 'Ω' : (result === 1 ? 'ohm' : 'ohms'); // eslint-disable-line no-nested-ternary

    let pow = '';
    if (result >= 1e9) {
      result /= 1e9;
      pow = short ? 'G' : 'giga';
    } else if (result >= 1e6) {
      result /= 1e6;
      pow = short ? 'M' : 'mega';
    } else if (result >= 1e3) {
      result /= 1e3;
      pow = short ? 'K' : 'kilo';
    }

    let tolerance = this.get('tolerance');
    if (this.get('bands') === 3) {
      tolerance = 20;
    }

    let string = `${result} ${pow}${unit} ± ${tolerance}%`;
    if (this.get('bands') === 6) {
      string += ` ${this.get('tempCoef')} ppm/K`;
    }
    return string;
  }

  static get significantDigits() {
    return significantDigits;
  }

  static get multipliers() {
    return multipliers;
  }

  static get tolerances() {
    return tolerances;
  }

  static get tempCoefs() {
    return tempCoefs;
  }

  static significantDigit(props) {
    return _find(significantDigits, props);
  }

  static multiplier(props) {
    return _find(multipliers, props);
  }

  static tolerance(props) {
    return _find(tolerances, props);
  }

  static tempCoef(props) {
    return _find(tempCoefs, props);
  }
}

export default Resistor;
