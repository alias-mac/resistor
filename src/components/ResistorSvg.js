/*
 * Copyright (c) 2016 Filipe Guerra
 * https://github.com/alias-mac/resistor
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */

import React from 'react';
import Resistor from '../Resistor';

function ResistorSvg(props) {

  const { model, ...other } = props;

  return (
    <svg {...other} width="100%" viewBox="0 0 400 100" preserveAspectRatio="xMinYMin meet">
      <defs>
        <rect id="edge" rx="10" ry="10" width="80" height="100" fill="#d6d6d6" />
        <rect id="stripe-wide" width="20" height="100%" />
        <rect id="stripe" y="10" width="20" height={100 - 20} />
      </defs>

      <rect x="0" y={50 - 12} width="100%" height="24" fill="#e8e8e8" />

      <rect x="60" y="10" width={400 - 120} height={100 - 20} fill="#d6d6d6" />

      <use xlinkHref="#edge" x="40" y="0" />
      <use xlinkHref="#edge" x={400 - 120} y="0" />

      <use
        id="stripe-1" xlinkHref="#stripe-wide" x="90"
        fill={Resistor.significantDigit({ value: model.get('digit1') }).color}
      />
      <use
        id="stripe-2" xlinkHref="#stripe" x="130"
        fill={Resistor.significantDigit({ value: model.get('digit2') }).color}
      />
      {(model.get('bands') === 5 || model.get('bands') === 6) &&
        <use
          id="stripe-3" xlinkHref="#stripe" x="160"
          fill={Resistor.significantDigit({ value: model.get('digit3') }).color}
        />
      }

      <use
        id="multiplier" xlinkHref="#stripe" x="190"
        fill={Resistor.multiplier({ value: model.get('multiplier') }).color}
      />

      {model.get('bands') !== 6 && model.get('bands') !== 3 &&
        <use
          id="tolerance" xlinkHref="#stripe-wide" x="290"
          fill={Resistor.tolerance({ value: model.get('tolerance') }).color}
        />
      }
      {model.get('bands') === 6 &&
        <use
          id="tolerance" xlinkHref="#stripe" x="250"
          fill={Resistor.tolerance({ value: model.get('tolerance') }).color}
        />
      }
      {model.get('bands') === 6 &&
        <use
          id="temperature" xlinkHref="#stripe-wide" x="290"
          fill={Resistor.tempCoef({ value: model.get('tempCoef') }).color}
        />
      }
    </svg>
  );
}

ResistorSvg.propTypes = {
  model: React.PropTypes.instanceOf(Resistor).isRequired,
};

export default ResistorSvg;
