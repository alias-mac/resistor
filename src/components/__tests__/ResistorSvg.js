/*
 * Copyright (c) 2016 Filipe Guerra
 * https://github.com/alias-mac/resistor
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */

import React from 'react';
import renderer from 'react-test-renderer';

import Resistor from '../../Resistor';
import ResistorSvg from '../ResistorSvg';

describe('ResistorSvg', function () {

  it('should render a 3-band resistor', function () {

    const resistor = new Resistor({
      bands: 3,
      digit1: 1,
      digit2: 0,
      multiplier: 0,
    });

    const component = renderer.create(<ResistorSvg model={resistor} />);

    expect(component.toJSON()).toMatchSnapshot();
  });

  it('should render a 4-band resistor', function () {

    const resistor = new Resistor({
      bands: 4,
      digit1: 1,
      digit2: 0,
      multiplier: 0,
      tolerance: 5,
    });

    const component = renderer.create(<ResistorSvg model={resistor} />);

    expect(component.toJSON()).toMatchSnapshot();
  });

  it('should render a 5-band resistor', function () {

    const resistor = new Resistor({
      bands: 5,
      digit1: 1,
      digit2: 0,
      digit3: 0,
      multiplier: 0,
      tolerance: 10,
    });

    const component = renderer.create(<ResistorSvg model={resistor} />);

    expect(component.toJSON()).toMatchSnapshot();
  });

  it('should render a 6-band resistor', function () {

    const resistor = new Resistor({
      bands: 6,
      digit1: 1,
      digit2: 0,
      digit3: 0,
      multiplier: 0,
      tolerance: 5,
      tempCoef: 15,
    });

    const component = renderer.create(<ResistorSvg model={resistor} />);

    expect(component.toJSON()).toMatchSnapshot();
  });

});
