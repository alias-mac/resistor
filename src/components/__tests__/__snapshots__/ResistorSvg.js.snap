exports[`ResistorSvg should render a 3-band resistor 1`] = `
<svg
  preserveAspectRatio="xMinYMin meet"
  viewBox="0 0 400 100"
  width="100%">
  <defs>
    <rect
      fill="#d6d6d6"
      height="100"
      id="edge"
      rx="10"
      ry="10"
      width="80" />
    <rect
      height="100%"
      id="stripe-wide"
      width="20" />
    <rect
      height="calc(100% - 20)"
      id="stripe"
      width="20"
      y="10" />
  </defs>
  <rect
    fill="#e8e8e8"
    height="24"
    width="100%"
    x="0"
    y="calc(50 - 12)" />
  <rect
    fill="#d6d6d6"
    height="calc(100% - 20)"
    width="calc(400 - 120)"
    x="60"
    y="10" />
  <use
    x="40"
    xlinkHref="#edge"
    y="0" />
  <use
    x="calc(100% - 120)"
    xlinkHref="#edge"
    y="0" />
  <use
    fill="brown"
    id="stripe-1"
    x="90"
    xlinkHref="#stripe-wide" />
  <use
    fill="black"
    id="stripe-2"
    x="130"
    xlinkHref="#stripe" />
  <use
    fill="black"
    id="multiplier"
    x="190"
    xlinkHref="#stripe" />
</svg>
`;

exports[`ResistorSvg should render a 4-band resistor 1`] = `
<svg
  preserveAspectRatio="xMinYMin meet"
  viewBox="0 0 400 100"
  width="100%">
  <defs>
    <rect
      fill="#d6d6d6"
      height="100"
      id="edge"
      rx="10"
      ry="10"
      width="80" />
    <rect
      height="100%"
      id="stripe-wide"
      width="20" />
    <rect
      height="calc(100% - 20)"
      id="stripe"
      width="20"
      y="10" />
  </defs>
  <rect
    fill="#e8e8e8"
    height="24"
    width="100%"
    x="0"
    y="calc(50 - 12)" />
  <rect
    fill="#d6d6d6"
    height="calc(100% - 20)"
    width="calc(400 - 120)"
    x="60"
    y="10" />
  <use
    x="40"
    xlinkHref="#edge"
    y="0" />
  <use
    x="calc(100% - 120)"
    xlinkHref="#edge"
    y="0" />
  <use
    fill="brown"
    id="stripe-1"
    x="90"
    xlinkHref="#stripe-wide" />
  <use
    fill="black"
    id="stripe-2"
    x="130"
    xlinkHref="#stripe" />
  <use
    fill="black"
    id="multiplier"
    x="190"
    xlinkHref="#stripe" />
  <use
    fill="gold"
    id="tolerance"
    x="290"
    xlinkHref="#stripe-wide" />
</svg>
`;

exports[`ResistorSvg should render a 5-band resistor 1`] = `
<svg
  preserveAspectRatio="xMinYMin meet"
  viewBox="0 0 400 100"
  width="100%">
  <defs>
    <rect
      fill="#d6d6d6"
      height="100"
      id="edge"
      rx="10"
      ry="10"
      width="80" />
    <rect
      height="100%"
      id="stripe-wide"
      width="20" />
    <rect
      height="calc(100% - 20)"
      id="stripe"
      width="20"
      y="10" />
  </defs>
  <rect
    fill="#e8e8e8"
    height="24"
    width="100%"
    x="0"
    y="calc(50 - 12)" />
  <rect
    fill="#d6d6d6"
    height="calc(100% - 20)"
    width="calc(400 - 120)"
    x="60"
    y="10" />
  <use
    x="40"
    xlinkHref="#edge"
    y="0" />
  <use
    x="calc(100% - 120)"
    xlinkHref="#edge"
    y="0" />
  <use
    fill="brown"
    id="stripe-1"
    x="90"
    xlinkHref="#stripe-wide" />
  <use
    fill="black"
    id="stripe-2"
    x="130"
    xlinkHref="#stripe" />
  <use
    fill="black"
    id="stripe-3"
    x="160"
    xlinkHref="#stripe" />
  <use
    fill="black"
    id="multiplier"
    x="190"
    xlinkHref="#stripe" />
  <use
    fill="silver"
    id="tolerance"
    x="290"
    xlinkHref="#stripe-wide" />
</svg>
`;

exports[`ResistorSvg should render a 6-band resistor 1`] = `
<svg
  preserveAspectRatio="xMinYMin meet"
  viewBox="0 0 400 100"
  width="100%">
  <defs>
    <rect
      fill="#d6d6d6"
      height="100"
      id="edge"
      rx="10"
      ry="10"
      width="80" />
    <rect
      height="100%"
      id="stripe-wide"
      width="20" />
    <rect
      height="calc(100% - 20)"
      id="stripe"
      width="20"
      y="10" />
  </defs>
  <rect
    fill="#e8e8e8"
    height="24"
    width="100%"
    x="0"
    y="calc(50 - 12)" />
  <rect
    fill="#d6d6d6"
    height="calc(100% - 20)"
    width="calc(400 - 120)"
    x="60"
    y="10" />
  <use
    x="40"
    xlinkHref="#edge"
    y="0" />
  <use
    x="calc(100% - 120)"
    xlinkHref="#edge"
    y="0" />
  <use
    fill="brown"
    id="stripe-1"
    x="90"
    xlinkHref="#stripe-wide" />
  <use
    fill="black"
    id="stripe-2"
    x="130"
    xlinkHref="#stripe" />
  <use
    fill="black"
    id="stripe-3"
    x="160"
    xlinkHref="#stripe" />
  <use
    fill="black"
    id="multiplier"
    x="190"
    xlinkHref="#stripe" />
  <use
    fill="gold"
    id="tolerance"
    x="250"
    xlinkHref="#stripe" />
  <use
    fill="orange"
    id="temperature"
    x="290"
    xlinkHref="#stripe-wide" />
</svg>
`;
